go/consensus/tendermint/apps/roothash/roothash.go

```diff
@@ -580,13 +526,49 @@ func (app *rootHashApplication) tryFinalizeBlock(
 	ctx *tmapi.Context,
 	rtState *roothashState.RuntimeState,
 	forced bool,
-) error {
+) (err error) {
+	defer func(previousTimeout int64) {
+		if err != nil {
+			return
+		}
+
+		// Do not re-arm the round timeout if the timeout has not changed.
+		nextTimeout := rtState.ExecutorPool.NextTimeout
+		if previousTimeout == nextTimeout {
+			return
+		}
+
+		state := roothashState.NewMutableState(ctx.State())
+		if previousTimeout != commitment.TimeoutNever {
+			if err = state.ClearRoundTimeout(ctx, rtState.Runtime.ID, previousTimeout); err != nil {
+				err = fmt.Errorf("failed to clear round timeout: %w", err)
+				return
+			}
+		}
+
+		switch nextTimeout {
+		case commitment.TimeoutNever:
+			// Only clear round timeout (already done).
+			ctx.Logger().Debug("disarming round timeout")
+		default:
+			// Set a different round timeout.
+			ctx.Logger().Debug("(re-)arming round timeout",
+				"height", ctx.BlockHeight(),
+				"next_timeout", nextTimeout,
+			)
+			if err = state.ScheduleRoundTimeout(ctx, rtState.Runtime.ID, nextTimeout); err != nil {
+				err = fmt.Errorf("failed to schedule round timeout: %w", err)
+				return
+			}
+		}
+	}(rtState.ExecutorPool.NextTimeout)
+
 	finalizedBlock := app.tryFinalizeExecutor(ctx, rtState, forced)
 	if finalizedBlock == nil {
 		return nil
 	}
 
-	if err := app.postProcessFinalizedBlock(ctx, rtState, finalizedBlock); err != nil {
+	if err = app.postProcessFinalizedBlock(ctx, rtState, finalizedBlock); err != nil {
```

Comment:
hm i don't think the `app.postProcessFinalizedBlock` can even ever return an error, maybe update the function definition.

Also the initial error check can then be skipped in the deffered function

Reply:
Changed.